### ISSUE TITLE
Using a private attribute in model_post_init of subclass

### ISSUE THREAD

**Issue by jrapin** [pydantic/pydantic#9098]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

Using/accessing/checking a private attribute in `model_post_init` of a subclass *before* calling `super().model_post_init` raises a cryptic `TypeError: 'NoneType' object is not subscriptable`.
My (erroneous?) expectation would be that the private attributes should already have been initialized when calling `model_post_init`

Traceback:
```python traceback
TypeError                                 Traceback (most recent call last)
Cell In[7], line 16
     13             self._priv = 13
     14         super().model_post_init(log__)  # no more bug if moved to the start
---> 16 obj = Cls()

    [... skipping hidden 1 frame]

Cell In[7], line 11, in Cls.model_post_init(self, log__)
     10 def model_post_init(self, log__: tp.Any) -> None:
---> 11     if self._priv == 12:  # TypeError: 'NoneType' object is not subscriptable
     12         # (same bug with getattr, hasattr)
     13         self._priv = 13
     14     super().model_post_init(log__)

File ~/Applications/miniconda3/envs/xxx/lib/python3.10/site-packages/pydantic/main.py:746, in BaseModel.__getattr__(self, item)
    742     return attribute.__get__(self, type(self))  # type: ignore
    744 try:
    745     # Note: self.__pydantic_private__ cannot be None if self.__private_attributes__ has items
--> 746     return self.__pydantic_private__[item]  # type: ignore
    747 except KeyError as exc:
    748     raise AttributeError(f'{type(self).__name__!r} object has no attribute {item!r}') from exc

TypeError: 'NoneType' object is not subscriptable
```

### Example Code

```Python
import typing as tp
import pydantic


class BaseCls(pydantic.BaseModel):
    _priv: tp.Any = 12


class Cls(BaseCls):
    def model_post_init(self, log__: tp.Any) -> None:
        if self._priv == 12:  # TypeError: 'NoneType' object is not subscriptable
            # (same bug with getattr, hasattr)
            self._priv = 13
        super().model_post_init(log__)  # no more bug if moved to the start

obj = Cls()
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.6.4
        pydantic-core version: 2.16.3
          pydantic-core build: profile=release pgo=true
                 install path: /Users/jrapin/Applications/miniconda3/envs/neuralset/lib/python3.10/site-packages/pydantic
               python version: 3.10.12 (main, Jul  5 2023, 15:02:25) [Clang 14.0.6 ]
                     platform: macOS-14.3-arm64-arm-64bit
             related packages: pyright-1.1.334 mypy-1.8.0 typing_extensions-4.7.1
                       commit: unknown
```

**Comment by sydney-runkle** [comment]
At a first glance, looks similar to https://github.com/pydantic/pydantic/issues/7293, which I thought we fixed...

**Comment by sydney-runkle** [comment]
@Viicos, any insight on what we should do here? I'm wondering why we don't have a check for `if private_attributes or base_private_attributes` here, but perhaps there's an obvious reason why that I'm missing: https://github.com/pydantic/pydantic/blob/426402af51602aaa773e283c4f12ae17a36c38b5/pydantic/_internal/_model_construction.py#L96

**Comment by jrapin** [comment]
> At a first glance, looks similar to #7293, which I thought we fixed...

Oh that one was from me as well :D it's different and it was fixed indeed (+ there's a unit test about it). I guess I'm a bit pushing the limits of pydantic private attributes sorry :s

**Comment by sydney-runkle** [comment]
@jrapin,

No worries at all - we appreciate the feedback + issue reports! We'll dig around for a fix!

**Comment by Viicos** [comment]
> I'm wondering why we don't have a check for `if private_attributes or base_private_attributes` here, but perhaps there's an obvious reason why that I'm missing:

This was simply missed in the original fix. Opened a PR

**Comment by sydney-runkle** [comment]
This fix will be released with 2.7 early next week :).
